llvm/lib/Target/RISCV/RISCVInstrInfoXCheri.td

```diff
@@ -1411,6 +1443,21 @@ defm : CheriExplicitStPat<store, SD_DDC, GPR, GPR>;
 defm : CheriExplicitStPat<store, SC_DDC_128, GPCR, GPR>;
 }
 
+let Predicates = [HasCheri, HasStdExtF],
+    AddedComplexity = -10 in
+def : Pat<(store FPR32:$rs2, GPR:$rs1),
```

Comment:
Should this be restricted to IsCapMode? I assume the default pattern will be picked first anyway, but it should not be needed for hybrid?

Reply:
The ones that were here used to be. However, the pattern _is_ legal, it’s just not the desired one, hence the AddedComplexity. This follows what I already did for loading/storing integers. It’s true there’s precedent for using NotFoo to guide this kind of thing though now (various bitmanip things use that scheme) so perhaps this is not a good approach, but I’d prefer everything be consistent, whatever that is.

Reply:
I'm fine with both approaches (with a slight preference towards preventing LLVM from ever generating it instead of using addedcomplexity).

However, I think it might be worth documenting the AddedComplexity with a comment since I find it less obvious than the NotCapMode/IsCapMode. Maybe something like:
```
// These patterns are legal in both modes, but we prefer the non-explicit address mode loads/stores since they have an immediate. Setting AddedComplexity = -10 ensures that LLVM prefers the standard loads/stores.
```

Reply:
I think I'll just follow up with adding CapMode predicates; when I wrote these there wasn't precedent in the backend for using predicates that way as nothing really overlapped, but now there is I'm happy to do that